### ISSUE TITLE
Allow disabling the Server: TwistedWeb header in the HTTP server

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7910]
|<img alt="Alex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| Alex reported|
|-|-|
|Trac ID|trac#7910|
|Type|enhancement|
|Created|2015-05-21 15:44:15Z|

Right now there's no API to disable sending this header, as far as I can tell.

<details><summary>Searchable metadata</summary>

```
trac-id__7910 7910
type__enhancement enhancement
reporter__Alex Alex
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__new new
resolution__None None
component__web web
keywords__ 
time__1432223055815976 1432223055815976
changetime__1432228351821470 1432228351821470
version__None None
owner__None None

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_1"></a>|@adiroiban commented|
|-|-|

Thanks for the but report.

Can you please suggest a possible API?

For my project I use a modified version of twisted.web.server.Request ...

what I wanted was to control the `set various default headers` block 
https://github.com/twisted/twisted/blob/trunk/twisted/web/server.py#L175

I have extracted that part into a method and implemented my own default headers.

```
    def _setDefaultHeaders(self):
        """
        Called to set headers to all responses.
        """
```

with such an API user can overwrite this method and add/delete headers

**Comment by zhuofan-16** [comment]
I think this issue is important. There is a security vulnerability posted when the version is disclosed.

**Comment by zhuofan-16** [comment]
Any updates regarding this just wondering?

**Comment by adiroiban** [comment]
My comment is the same: Can you please suggest a possible API?

How do you expect to be able to disalbe this?

-----------

Also, you can always explicitly hide the header from the requets,

I agree that this is just a workaround.

```python
from twisted.internet import reactor, endpoints
from twisted.web import server
from twisted.web.resource import Resource

class HideHeader(Resource):
    isLeaf = True
    def render_GET(self, request):
        request.setHeader('server', '')
        return b'NO TWISTED SERVER HEADER\n'



endpoint = endpoints.TCP4ServerEndpoint(reactor, 8080)
endpoint.listen(server.Site(HideHeader()))
reactor.run()
```


```shell
$ curl -v http://localhost:8080
*   Trying 127.0.0.1:8080...
* Connected to localhost (127.0.0.1) port 8080 (#0)
> GET / HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.81.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Server: 
< Date: Mon, 11 Sep 2023 10:40:38 GMT
< Content-Length: 25
< Content-Type: text/html
< 
NO TWISTED SERVER HEADER
```

**Comment by zhuofan-16** [comment]
> My comment is the same: Can you please suggest a possible API?
> 
> How do you expect to be able to disalbe this?
> 
> Also, you can always explicitly hide the header from the requets,
> 
> I agree that this is just a workaround.
> 
> ```python
> from twisted.internet import reactor, endpoints
> from twisted.web import server
> from twisted.web.resource import Resource
> 
> class HideHeader(Resource):
>     isLeaf = True
>     def render_GET(self, request):
>         request.setHeader('server', '')
>         return b'NO TWISTED SERVER HEADER\n'
> 
> 
> 
> endpoint = endpoints.TCP4ServerEndpoint(reactor, 8080)
> endpoint.listen(server.Site(HideHeader()))
> reactor.run()
> ```
> 
> ```shell
> $ curl -v http://localhost:8080
> *   Trying 127.0.0.1:8080...
> * Connected to localhost (127.0.0.1) port 8080 (#0)
> > GET / HTTP/1.1
> > Host: localhost:8080
> > User-Agent: curl/7.81.0
> > Accept: */*
> > 
> * Mark bundle as not supporting multiuse
> < HTTP/1.1 200 OK
> < Server: 
> < Date: Mon, 11 Sep 2023 10:40:38 GMT
> < Content-Length: 25
> < Content-Type: text/html
> < 
> NO TWISTED SERVER HEADER
> ```

Currently, I am using  flask as the wsgi application. Running twisted web using 

`export PYTHONPATH=${PYTHONPATH}:${PWD} # exports python path
twistd  web -n --port tcp:5000 --wsgi path-to-your-app-root --logfile log.txt`

Saw in twistd --help that there is a --add-header para, would it be possible to have something like --remove-header or --edit-header

**Comment by zhuofan-16** [comment]
Or, another approach maybe check if the server field exist in the response header first? If yes, then dont add/modify, else add the server field.

**Comment by adiroiban** [comment]
I think that if you use `--add-header` in `twistd` you should be able to "Edit" the existing `Server` header.

Have you tried  that ? Do you get 2 separate `Server` headers?

**Comment by zhuofan-16** [comment]
> I think that if you use `--add-header` in `twistd` you should be able to "Edit" the existing `Server` header.
> 
> Have you tried that ? Do you get 2 separate `Server` headers?

Ya I tried that , you will get 2 separate server header

![MicrosoftTeams-image](https://github.com/twisted/twisted/assets/59607914/f8b4dcc2-a40f-41bc-8e88-31cae7dfac20)

**Comment by adiroiban** [comment]
thanks for the update.

Ok. then let's focus on the `twistd web` functionality

I would say that if you do `twistd web --add-header="Server: value"` it  should only generate one header.

if you call `twistd web --add-header="Server: value" --add-header="Server: other-value"` it should generate 2 headers. 


Even if you call ``twistd web --add-header="Server: value" --add-header="Server: value"` it should still generate 2 headers.

-------------

This looks like a reasonable UI.

Feel free to submit a PR to implement these changed and I will try to review them.

Regards

**Comment by exarkun** [comment]
> This looks like a reasonable UI.

Obviously, it excludes the possibility of configuring a server to leave the default response-header in place and add additional custom values.  A good UI would allow configuring all legal values for a response.

**Comment by adiroiban** [comment]
The default headers are

* `Date`
* `Server`
* `Content-Length`
* `Content-Type`
* and maybe a few others

Is there a use case in which you would want to add another of these headers and still keep the  default value?

I think that adding `--remove-header` or `--edit-header` is not the best UI/UX/CLI

Changing the semantic/behaviour of `--add-header` might not be the best option.

I am happy to have this implemented in any other way.

Regards
